### PR TITLE
Cryptostuff

### PR DESCRIPTION
**Pull request by JHanselman**
Weil pairing, supersingular tests, fixed multiplication_by_map, projective coordinates

### CODE REVIEW

File: src/EllCrv/Finite.jl [hunk 1]
```diff
@@ -816,3 +817,132 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
...
+
+#Following Identifying supersingular elliptic curves - Andrew V. Sutherland
+@doc Markdown.doc"""
+    supersingular_monte_carlo(E::EllCrv{T}) where T <: FinFieldElem
+Uses a probabilistic algorithm to test whether E is supersingular or not.
+If the function returns false, the curve is proven to be ordinary.
+If the function returns true, there is a high chance the curve is supersingular,
+but the result hasn't been proven.
+"""
+function supersingular_monte_carlo(E::EllCrv{T}) where T <: FinFieldElem
```
Comment (thofma) [inline src/EllCrv/Finite.jl:895]:
```suggestion
function is_probable_supersingular(E::EllCrv{T}) where T <: FinFieldElem
```
(To be consistent with `is_probable_prime`).

File: test/EllCrv/Finite.jl [hunk 2]
```diff
@@ -95,4 +95,33 @@
...
+    @inferred supersingular_monte_carlo(E) == true
+    
+    E = elliptic_curve_from_j_invariant(K(170))
+    @test @inferred is_ordinary(E) == true
+    
+    K = GF(193, 3)
+    a = gen(K)
+    E = elliptic_curve_from_j_invariant(a)
+    @test @inferred is_supersingular(E) == false
+    @inferred supersingular_monte_carlo(E) == false
```
Comment (thofma) [inline test/EllCrv/Finite.jl:124]:
```suggestion
    @inferred supersingular_monte_carlo(E)
```

File: src/EllCrv/Finite.jl (hunk 1 above)
Comment (thofma) [reply src/EllCrv/Finite.jl:895]: Needs to be then adjusted in the tests.

File: src/EllCrv/Finite.jl [hunk 3]
```diff
@@ -816,3 +817,161 @@ function trace_of_frobenius(E::EllCrv{T}, n::Int) where T<:FinFieldElem
...
+  if degree(K) == 1
+    degj = 1
+  else
+    degj = degree(minpoly(j))
+  end
+  
+  if degj == 1
+    return monte_carlo_test(E, p+1)
+  elseif degj == 2
+    return monte_carlo_test(E, p+1) || monte_carlo_test(E, p+1)
```
Comment (thofma) [inline src/EllCrv/Finite.jl:934]: Stupid question, but should the second be `p-1`?

Comment (JHanselman) [reply src/EllCrv/Finite.jl]: It should be. I changed it.

**Pull request merged**